cache/lru_cache.cc

```diff
@@ -430,30 +433,62 @@ void LRUCacheShard::Promote(LRUHandle* e) {
 
   assert(secondary_handle->IsReady());
   e->SetIncomplete(false);
-  e->SetInCache(true);
   e->value = secondary_handle->Value();
-  e->CalcTotalCharge(secondary_handle->Size(), metadata_charge_policy_);
+  size_t secondary_handle_size = secondary_handle->Size();
+  e->CalcTotalCharge(secondary_handle_size, metadata_charge_policy_);
   delete secondary_handle;
 
   // This call could fail if the cache is over capacity and
   // strict_capacity_limit_ is true. In such a case, we don't want
   // InsertItem() to free the handle, since the item is already in memory
-  // and the caller will most likely just read from disk if we erase it here.
+  // and the caller will most likely just read from disk if we erase it
+  // here.
   if (e->value) {
-    Cache::Handle* handle = reinterpret_cast<Cache::Handle*>(e);
-    Status s = InsertItem(e, &handle, /*free_handle_on_fail=*/false);
+    bool insert_dummy_handle{false};
+    {
+      // Insert a dummy handle and return a standalone handle to caller
+      // when secondary_cache_ is CompressedSecondaryCache and e is a
+      // standalone handle.
+      if (use_compressed_secondary_cache_ && e->IsStandalone()) {
```

Comment:
we have a choice here. We can choose to insert the actual block here to allow other readers to fetch it, and only destroy it when the last reference is de-refed. I agree that it might be too complicated with little gain. Perhaps document the reasoning here.